### ISSUE TITLE
Project structure re-organizing.

### ISSUE THREAD

**Issue by PJB3005** [space-wizards/RobustToolbox#176]

**Comment by PJB3005** [comment]
You have to be fucking kidding me who designed the form for GitHub mobile's new issue page
Hang on while I write up a proper issue God damnit.

**Comment by PJB3005** [comment]
Ok proper issue time:

One thing I've been wondering about is why exactly is the project structure the way it is.

Currently we have the following:
* `SS14.Shared.IoC`
* `SS14.Shared`
* `SS14.Shared.GameObjects`
* **NO** `SS14.Shared.Interfaces`
* `SS14.Client`
* `SS14.Client.Graphics`
* `SS14.Client.Services`
* `SS14.Client.Interfaces`
* `SS14.Client.GameObjects`
* `SS14.Server`
* `SS14.Server.Services`
* `SS14.Server.Interfaces`
* `SS14.Server.GameObjects`

This is God damn dumb for a couple of reasons.

First of all, the "main" executables do basically nothing except provide one file that could have just been in `SS14.*.Services`.

Following up on the above: Services is just "all code, except game objects and graphics". I understand that historically Services was the only assemblies loaded into IoC, but this design limitation just made everything painful and cause hard to follow code everywhere. So why not move all of services into the main executables?

Game objects is just... Why? I guess it's there because it was originally intended to be the place the content would reside but that's everything I can think of. No reason for this not to be part of services and thus the main executable.

About interfaces. I think the system is great and works well, but having a separate assembly just feels strange. Shared doesn't have an interfaces module at all for example and a lot is done by putting implementations next to interfaces (I think at least, it's a mess due to lack of IoC). This is kind of a maybe point for mr: should we move interfaces into the main executable too?

So I propose the following structure:
* `SS14.Shared.IoC`
* `SS14.Shared`
* `SS14.Shared.Interfaces` *maybe merge with above*
* `SS14.Client`
* `SS14.Client.Graphics`
* `SS14.Client.Interfaces` *maybe*
* `SS14.Server`
* `SS14.Server.Interfaces` *maybe*

**Comment by elwray** [comment]
@PJB3005 what reasons to use "SS14.Shared.IoC"? As I known this is similar to Ninject, SimpleInjector and other DI libraries?
I think best solution to remove this project and use SimpleInjector, for example.

**Comment by PJB3005** [comment]
Honestly that's probably a good idea. Didn't even cross my mind.

**Comment by ZoldorfTheWizard** [comment]
Makes sense to me.  SS14.Client.Graphics will be refactored into SS14.RobustEngine
Unless i make it a nuget package

**Comment by PJB3005** [comment]
Ok so revised plan (`/` means it's in a namespace, but the same assembly)

* `SS14.Shared`
* `SS14.Shared/IoC`
* `SS14.Shared/Interfaces`
* `SS14.Client`
* `SS14.Client/Interfaces`
* `SS14.RobustEngine`, separated maybe.
* `SS14.Server`
* `SS14.Server/Interfaces`

Feedback please.

**Comment by PJB3005** [comment]
Issue with SimpleInjector:

https://simpleinjector.readthedocs.io/en/latest/decisions.html#the-container-is-locked-after-the-first-call-to-resolve

Into the trash it goes.
